packages/contracts/contracts/TransactionManager.sol

```diff
@@ -503,11 +499,22 @@ contract TransactionManager is ReentrancyGuard, ProposedOwnable, ITransactionMan
       // on the sending chain
       require(msg.sender == txData.router, "#F:016");
 
+      // Validate the user has signed
+      require(recoverFulfillSignature(txData.transactionId, relayerFee, txData.receivingChainId, txData.receivingChainTxManagerAddress, signature) == txData.user, "#F:022");
+
       // Complete tx to router for original sending amount
       routerBalances[txData.router][txData.sendingAssetId] += txData.amount;
 
     } else {
-      (success, returnData) = _receivingChainFulfill(txData, relayerFee, callData);
+      // Validate the user has signed, using domain of contract
+      require(recoverFulfillSignature(txData.transactionId, relayerFee, _chainId, address(this), signature) == txData.user, "#F:022");
+
+      // Sanity check: fee <= amount. Allow `=` in case of only 
+      // wanting to execute 0-value crosschain tx, so only providing 
+      // the fee amount
+      require(relayerFee <= txData.amount, "#F:023");
```

Comment:
is it possible to execute a 0-value tx where the relayer gets the entire amount? wouldn't there be other participants that the fee would be paid out to, thereby detracting more portions from the amount? what happens if there are multiple relayers exactly? if this relayer is taking the full amount as the fee, how does the next relayer get their fee?

Reply:
This is on the receiver side -- the router fee is already implicitly included because the amount it is prepared with on the receiving chain is less than the amount the user prepared with on the sending chain. That difference is the router fee.

The relayer fee is for someone who is submitting the transactions to chain. There can only be one person who claims the fee, since the transaction can only be executed once. The relayer could get the entire amount from the transaction, and this would make sense in the case where you want to execute calldata on a chain you don't have gas on. You wouldn't want your funds there, and including a relayer fee creates an incentive for someone to submit the tx on your behalf.